tests/erlang_tests/test_utf8_atoms.erl

```diff
@@ -0,0 +1,232 @@
+    comp(
+        erlang:list_to_atom(get_list(l1s_missing)),
+        erlang:binary_to_atom(get_binary(l1s_missing), utf8)
+    ) +
+        comp(
+            erlang:list_to_atom(get_list(jp_mixed_missing)),
+            erlang:binary_to_atom(get_binary(jp_mixed_missing), utf8)
+        ) * 2 +
+        comp(
+            erlang:list_to_atom(get_list(l1s_missing)),
+            erlang:binary_to_atom(get_binary(l1s_missing), unicode)
+        ) * 4 +
+        comp(
+            erlang:list_to_atom(get_list(jp_mixed_missing)),
+            erlang:binary_to_atom(get_binary(jp_mixed_missing), unicode)
+        ) * 8 +
+        comp_opt(
+            fun() -> erlang:list_to_atom(get_list(l1s_missing)) end,
+            fun() -> erlang:binary_to_atom(get_binary(l1s_missing)) end
+        ) * 16 +
+        comp_opt(
+            fun() -> erlang:list_to_atom(get_list(jp_mixed_missing)) end,
+            fun() -> erlang:binary_to_atom(get_binary(jp_mixed_missing)) end
+        ) * 32.
+
+test_invalid_bins() ->
+    test_from_u8bincatch(invalid1) +
+        test_from_u8bincatch(invalid2) * 2 +
+        test_from_u8bincatch(invalid2) * 4.
+
+test_to_list(Id) ->
+    case ?MODULE:conv(list, x, ?MODULE:get_atom(Id)) == ?MODULE:get_list(Id) of
+        true ->
+            1;
+        false ->
+            erlang:display({list, Id}),
+            0
+    end.
+
+test_to_l1bin(Id) ->
+    case ?MODULE:conv(binary, l1, ?MODULE:get_atom(Id)) == ?MODULE:get_binary(Id) of
+        true ->
+            1;
+        false ->
+            erlang:display({l1bin, Id}),
+            0
+    end.
+
+test_to_l1bincatch(Id) ->
+    try ?MODULE:conv(binary, l1, ?MODULE:get_atom(Id)) of
+        _X ->
+            erlang:display({err, Id}),
+            0
+    catch
+        error:badarg ->
+            1
+    end.
+
+test_to_u8bin(Id) ->
+    case ?MODULE:conv(binary, u8, ?MODULE:get_atom(Id)) == ?MODULE:get_binary(Id) of
+        true ->
+            1;
+        false ->
+            erlang:display({u8bin, Id}),
+            0
+    end.
+
+test_from_list(Id) ->
+    case ?MODULE:conv2(list, x, ?MODULE:get_list(Id)) == ?MODULE:get_atom(Id) of
+        true ->
+            1;
+        false ->
+            erlang:display({flist, Id}),
+            0
+    end.
+
+test_from_l1bin(Id) ->
+    case ?MODULE:conv2(binary, l1, ?MODULE:get_binary(Id)) == ?MODULE:get_atom(Id) of
+        true ->
+            1;
+        false ->
+            erlang:display({fl1bin, Id}),
+            0
+    end.
+
+test_from_u8bin(Id) ->
+    case ?MODULE:conv2(binary, u8, ?MODULE:get_binary(Id)) == ?MODULE:get_atom(Id) of
+        true ->
+            1;
+        false ->
+            erlang:display({fu8bin, Id}),
+            0
+    end.
+
+test_from_u8bincatch(Id) ->
+    try ?MODULE:conv2(binary, u8, ?MODULE:get_binary(Id)) of
+        _X ->
+            erlang:display({u8err, Id}),
+            0
+    catch
+        error:badarg ->
+            1
+    end.
+
+conv(list, _Fmt, Atom) ->
```

Comment:
I find these two functions (conv/3, conv2/3) are less readable than inline functions IMHO.
Are you afraid of some kind of compiler optimization?

I would say it's more readable to use an `id/1` function to avoid them.

Reply:
Compiler is getting smarter and smarter, I'm not sure id trick will work on longer term so I'm future-proofing it a bit.